core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -513,11 +469,116 @@ void toSerializable(DistributedStorage storage)
+        scanner = storage.scan(Scan.newBuilder(scan).limit(0).build());
+      }
+
+      Iterator<Entry<Key, TransactionResult>> resultsIterator = results.entrySet().iterator();
+
+      Optional<Result> latestResult = scanner.one();
+
+      Entry<Key, TransactionResult> resultsEntry = null;
+      if (resultsIterator.hasNext()) {
+        resultsEntry = resultsIterator.next();
+      }
+
+      while (latestResult.isPresent() && resultsEntry != null) {
+        TransactionResult latestTransactionResult = new TransactionResult(latestResult.get());
+        Key key = new Key(scan, latestTransactionResult);
+
+        // Ignore records that this transaction has prepared (and that are in the write set or
+        // delete set)
+        if (latestTransactionResult.getId() != null && latestTransactionResult.getId().equals(id)) {
+          latestResult = scanner.one();
+
+          if (resultsEntry.getKey().equals(key)) {
+            // The record is updated by this transaction
+
+            if (resultsIterator.hasNext()) {
+              resultsEntry = resultsIterator.next();
+            } else {
+              resultsEntry = null;
+            }
+          } else {
+            // The record is inserted/deleted by this transaction
+          }
+
+          continue;
+        }
+
+        if (!resultsEntry.getKey().equals(key)) {
+          // The record is inserted/deleted by another transaction
+          throwExceptionDueToAntiDependency();
+        }
+
+        if (isChanged(latestTransactionResult, resultsEntry.getValue())) {
+          // The record is updated by another transaction
+          throwExceptionDueToAntiDependency();
+        }
+
+        latestResult = scanner.one();
+
+        if (resultsIterator.hasNext()) {
+          resultsEntry = resultsIterator.next();
+        } else {
+          resultsEntry = null;
+        }
+      }
+
+      if (resultsEntry != null) {
+        // Some of the records are deleted by another transaction
+        throwExceptionDueToAntiDependency();
+      }
+
+      if (!scanWithLimit) {
+        while (latestResult.isPresent()) {
+          TransactionResult latestTransactionResult = new TransactionResult(latestResult.get());
+
+          // Ignore records that this transaction has prepared (and that are in the write set or
+          // delete set)
+          if (latestTransactionResult.getId() != null
+              && latestTransactionResult.getId().equals(id)) {
+            // The record is inserted/deleted by this transaction
+
+            latestResult = scanner.one();
+          } else {
+            // The record is inserted by another transaction
+            throwExceptionDueToAntiDependency();
+          }
+        }
+      }
+    } finally {
+      if (scanner != null) {
+        try {
+          scanner.close();
+        } catch (IOException e) {
+          logger.warn("Failed to close the scanner", e);
+        }
+      }
+    }
+  }
```

Comment:
To address issue 2, I’ve rewritten the validation logic for the scan set.

The purpose of this validation is to ensure that the scan results have not been modified by other transactions.

Key considerations for the validation:
- Since the validation is performed after the prepare-record phase, the scan might include prepared records if the transaction has performed operations that affect the scan result.
  - In such cases, those prepared records can be safely ignored.
- For scan operations with a limit, we only need to validate the records up to the number of records that were returned during the execution phase.